### PR TITLE
Upgrade CS

### PR DESCRIPTION
**Pull request by Forpee**
This PR aims to address #331 

- Ported `bellpepper-core` (& some parts of `bellpepper`) + Poseidon implementation from @argumentcomputer.

~~Remaining items:~~

 ~~There’s some unused code from the bellpepper and Poseidon port. Should we remove it now, or do you think it might be useful to keep for potential future needs?~~

### CODE REVIEW

**Comment by srinathsetty** [comment]
Hi @Forpee! thanks! 

Yes, it is preferred to have a minimal implementation of the required methods. Please attribute the original source if they are taken from an existing library.

File: benches/compressed-snark.rs [hunk 1]
```diff
@@ -1,9 +1,9 @@
 #![allow(non_snake_case)]
 
-use bellpepper_core::{num::AllocatedNum, ConstraintSystem, SynthesisError};
 use core::marker::PhantomData;
 use criterion::{measurement::WallTime, *};
 use ff::PrimeField;
+use nova_snark::frontend::{num::AllocatedNum, ConstraintSystem, SynthesisError};
```
Comment (srinathsetty) [inline benches/compressed-snark.rs:6]: Nit: this import and the next line's import can be merged into one. Can we please apply this everywhere?

**Comment by srinathsetty** [comment]
For poseidon code, it seems like we are not using a bunch of code. For example, in src/lib.rs, replace `pub mod provider` with `mod provider` (this makes clippy complain about unused code). I get the following unused code. Can we please trim code to minimize unused code?

error: variants `FullBuffer`, `IndexOutOfBounds`, `GpuError`, and `Other` are never constructed
   --> src/provider/poseidon/mod.rs:98:3
    |
96  | pub enum PoseidonError {
    |          ------------- variants in this enum
97  |   /// The allowed num...
98  |   FullBuffer,
    |   ^^^^^^^^^^
99  |   /// Attempt to reference an index element that is out of bounds
100 |   IndexOutOfBounds,
    |   ^^^^^^^^^^^^^^^^
101 |   /// GPU error
102 |   GpuError(String),
    |   ^^^^^^^^
103 |   /// Other error
104 |   Other(String),
    |   ^^^^^
    |
    = note: `PoseidonError` has derived impls for the traits `Clone` and `Debug`, but these are intentionally ignored during dead code analysis
note: the lint level is defined here
   --> src/lib.rs:4:3
    |
4   |   unused,
    |   ^^^^^^
    = note: `#[deny(dead_code)]` implied by `#[deny(unused)]`

error: methods `is_allocated` and `is_num` are never used
  --> src/provider/poseidon/circuit2.rs:31:16
   |
29 | impl<Scalar: PrimeField> Elt<Scalar> {
   | ------------------------------------ methods in this implementation
30 |   /// Check if the Elt is allocated.
31 |   pub const fn is_allocated(&self) -> bool {
   |                ^^^^^^^^^^^^
...
36 |   pub const fn is_num(&self) -> bool {
   |                ^^^^^^

error: variants `Correct` and `OptimizedDynamic` are never constructed
   --> src/provider/poseidon/poseidon_inner.rs:119:3
    |
116 | pub enum HashMode {
    |          -------- variants in this enum
...
119 |   Correct,
    |   ^^^^^^^
...
122 |   OptimizedDynamic,
    |   ^^^^^^^^^^^^^^^^
    |
    = note: `HashMode` has derived impls for the traits `Clone` and `Debug`, but these are intentionally ignored during dead code analysis

error: associated items `new_constant_length` and `with_length` are never used
   --> src/provider/poseidon/poseidon_inner.rs:150:10
    |
130 | / impl<F, A> PoseidonConstants<F, A>
131 | | where
132 | |   F: PrimeField,
133 | |   A: Arity<F>,
    | |______________- associated items in this implementation
...
150 |     pub fn new_constant_length(length: usize) -> Self {
    |            ^^^^^^^^^^^^^^^^^^^
...
157 |     pub fn with_length(&self, length: usize) -> Self {
    |            ^^^^^^^^^^^

error: associated items `new_with_preimage`, `set_preimage`, `reset`, and `input` are never used
   --> src/provider/poseidon/poseidon_inner.rs:298:10
    |
268 | / impl<'a, F, A> Poseidon<'a, F, A>
269 | | where
270 | |   F: PrimeField,
271 | |   A: Arity<F>,
    | |______________- associated items in this implementation
...
298 |     pub fn new_with_preimage(preimage: &[F], constants: &'a PoseidonConstan...
    |            ^^^^^^^^^^^^^^^^^
...
346 |     pub fn set_preimage(&mut self, preimage: &[F]) {
    |            ^^^^^^^^^^^^
...
353 |     pub fn reset(&mut self) {
    |            ^^^^^
...
368 |     pub fn input(&mut self, element: F) -> Result<usize, PoseidonError> {
    |            ^^^^^

error: methods `reset` and `is_squeeze` are never used
   --> src/provider/poseidon/sponge/api.rs:106:16
    |
104 | impl SpongeOp {
    | ------------- methods in this implementation
105 |   /// Reset the SpongeOp
106 |   pub const fn reset(&self) -> Self {
    |                ^^^^^
...
126 |   pub const fn is_squeeze(&self) -> bool {
    |                ^^^^^^^^^^

error: variant `Duplex` is never constructed
  --> src/provider/poseidon/sponge/vanilla.rs:38:3
   |
34 | pub enum Mode {
   |          ---- variant in this enum
...
38 |   Duplex,
   |   ^^^^^^
   |
   = note: `Mode` has a derived impl for the trait `Clone`, but this is intentionally ignored during dead code analysis

error: multiple associated items are never used
   --> src/provider/poseidon/sponge/vanilla.rs:77:6
    |
62  | pub trait SpongeTrait<'a, F: PrimeField, A: Arity<F>>
    |           ----------- associated items in this trait
...
77  |   fn simplex_constants(size: usize) -> PoseidonConstants<F, A> {
    |      ^^^^^^^^^^^^^^^^^
...
82  |   fn duplex_constants() -> PoseidonConstants<F, A> {
    |      ^^^^^^^^^^^^^^^^
...
121 |   fn make_elt(&self, val: F, acc: &mut Self::Acc) -> Self::Elt;
    |      ^^^^^^^^
...
147 |   fn is_squeezing(&self) -> bool {
    |      ^^^^^^^^^^^^
...
192 |   fn is_exhausted(&self) -> bool {
    |      ^^^^^^^^^^^^
...
333 |   fn absorb_elements(
    |      ^^^^^^^^^^^^^^^
...
345 |   fn squeeze_elements(&mut self, count: usize, acc: &mut Self::Acc) -> Vec<...
    |      ^^^^^^^^^^^^^^^^

File: benches/compressed-snark.rs (hunk 1 above)
Comment (Forpee) [reply benches/compressed-snark.rs:6]: will do, thanks!

File: src/frontend/constraint_system.rs [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+    false
+  }
+
+  /// Extend concatenates thew  `other` constraint systems to the receiver, modifying the receiver, whose
+  /// inputs, allocated variables, and constraints will precede those of the `other` constraint system.
+  /// The primary use case for this is parallel synthesis of circuits which can be decomposed into
+  /// entirely independent sub-circuits. Each can be synthesized in its own thread, then the
+  /// original `ConstraintSystem` can be extended with each, in the same order they would have
+  /// been synthesized sequentially.
+  fn extend(&mut self, _other: &Self) {
```
Comment (srinathsetty) [inline src/frontend/constraint_system.rs:142]: Nit: This method seems unimplemented. Is it worth cutting out unimplemented methods to keep the codebase simple?

File: src/frontend/constraint_system.rs [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+  }
+
+  /// Determines if the current `ConstraintSystem` instance is a witness generator.
+  /// ConstraintSystems that are witness generators need not assemble the actual constraints. Rather, they exist only
+  /// to efficiently create a witness.
+  ///
+  /// # Returns
+  ///
+  /// * `false` - By default, a `ConstraintSystem` is not a witness generator.
+  fn is_witness_generator(&self) -> bool {
```
Comment (srinathsetty) [inline src/frontend/constraint_system.rs:155]: Do we use this method anywhere?

File: src/frontend/util_cs/test_cs.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+  /// Check if the constraint system is satisfied.
+  pub fn is_satisfied(&self) -> bool {
+    match self.which_is_unsatisfied() {
+      Some(b) => {
+        println!("fail: {:?}", b);
+        false
+      }
+      None => true,
+    }
+    // self.which_is_unsatisfied().is_none()
```
Comment (srinathsetty) [inline src/frontend/util_cs/test_cs.rs:119]: Nit: remove comment?

File: src/frontend/constraint_system.rs (hunk 2 above)
Comment (srinathsetty) [reply src/frontend/constraint_system.rs:142]: It seems like there are some places this actually gets used. Will look more.

File: src/frontend/constraint_system.rs (hunk 3 above)
Comment (Forpee) [reply src/frontend/constraint_system.rs:155]: Yes, most notably in the Poseidon sponge circuit's `permute_state` method.

https://github.com/Forpee/Nova/blob/54cfd7d3bceed09d141e1c78bd20f7e8a4413c33/src/provider/poseidon/sponge/circuit.rs#L127

**Comment by Forpee** [comment]
Hi @srinathsetty,

I can address the clippy errors, but it seems these are issues that didn't previously fail (e.g., #needless_lifetimes and #empty_line_after_doc_comments). For instance, the latter appears in src/spartan/macros.rs, a file I didn't modify.

Fixing the #needless_lifetimes warnings would require changes to files unrelated to this PR, such as src/provider/hyperkzg.rs, src/provider/pedersen.rs, src/circuit.rs, src/r1cs/sparse.rs, and others. Let me know how you'd like to proceed!

**Comment by srinathsetty** [comment]
I’m making a separate PR with fixes. Thanks!

**Comment by srinathsetty** [comment]
@Forpee clippy warnings are now addressed. Please rebase with main.

**Comment by Forpee** [comment]
Okay, rebased.

File: src/provider/poseidon/round_constants.rs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    while !new_bit {
+      let _new_bit = self.generate_new_bit();
+      new_bit = self.generate_new_bit();
+    }
+    new_bit = self.generate_new_bit();
+    Some(new_bit)
+  }
+}
+
+#[allow(dead_code)]
```
Comment (srinathsetty) [inline src/provider/poseidon/round_constants.rs:169]: question: is this code needed? if not, shall we cut?

File: src/provider/poseidon/round_constants.rs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+const fn bool_to_u8(bit: bool, offset: usize) -> u8 {
+  if bit {
+    1u8 << offset
+  } else {
+    0u8
+  }
+}
+
+/// Converts a slice of bools into their byte representation, in little endian.
+#[allow(dead_code)]
```
Comment (srinathsetty) [inline src/provider/poseidon/round_constants.rs:180]: same question: is this code needed?

File: src/provider/poseidon/poseidon_inner.rs [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+
+use super::{
+  hash_type::HashType,
+  matrix::{left_apply_matrix, Matrix},
+  mds::{MdsMatrices, SparseMatrix},
+  quintic_s_box, round_constants, round_numbers, Strength, DEFAULT_STRENGTH,
+};
+
+/// Available arities for the Poseidon hasher.
+pub trait Arity<T>: ArrayLength {
```
Comment (srinathsetty) [inline src/provider/poseidon/poseidon_inner.rs:21]: Since Nova codebase only uses the Sponge type, do we need support for different Arity?

Comment (Forpee) [reply src/provider/poseidon/poseidon_inner.rs:21]: I think we can remove the "`Arity` option", and just hard code whatever `Arity` the `Nova` codebase uses?

Comment (srinathsetty) [reply src/provider/poseidon/poseidon_inner.rs:21]: Nova codebase does not explicitly set the arity and we only use the sponge APIs. Do you know how the sponge code selects arity?

Comment (Forpee) [reply src/provider/poseidon/poseidon_inner.rs:21]: From what I can see, passing `U24` as the second type parameter to the `Sponge`'s `api_constants` method sets the "arity" of the sponge.

https://github.com/microsoft/Nova/blob/8136faa864874440fc44cbd373d4b4de813338bd/src/provider/poseidon.rs#L30

**Pull request merged**